### PR TITLE
Expand the scope of useless-expression (B018)

### LINKED ISSUE #3453: [bug?] B018 `useless-expression` does not work
None of the following trigger `B018` `useless-expression` for me, when I would expect them all to:

```
"foo" + "bar"  # should trigger B018

"foo"  # should trigger B018

object().__class__  # should trigger B018
```

I might be missing something about what this rule is supposed to do though.

### PR DESCRIPTION
**Pull request by charliermarsh**
Closes #3453.

### CODE REVIEW

**Comment by charliermarsh** [comment]
I wanna see how this affects the ecosystem CI check, this might be too broad.

**Comment by charliermarsh** [comment]
This is currently flagging statements like `create_entry_group >> delete_entry_group` which are effect-ful.

**Comment by charliermarsh** [comment]
We probably need to omit all `BinOp` due to operator overloading. I suppose we could flag `BinOp` where the types are constant.

**Comment by charliermarsh** [comment]
I've been going back and forth on how aggressive to make this check. For example, should it flag expressions like:

```py
x  # Name
x.y  # Attribute
[x for x in y]  # ListComp
```

Right now it does (though bugbear doesn't pick up any of them). `Name` seems like an obvious thing to pick up. But `Attribute` could be effectful though if you have a custom getter.

@smackesey - Any opinion here? I'm guessing you want to err on the side of being closer to Pylint, which would flag all of these :)

**Comment by smackesey** [comment]
IMO they should all be flagged. I realize "Attribute" and "ListComp" _could_ be effectful, but there's really no good reason I can think of to write code where you rely on these unassigned expressions for side effects.  They are much more likely to be unintended when they occur.

File: crates/ruff/src/rules/flake8_bugbear/rules/useless_expression.rs [hunk 1]
```diff
@@ -1,41 +1,77 @@
...
-                    }
-                },
-                _ => {}
+pub fn useless_expression(checker: &mut Checker, value: &Expr) {
+    // Ignore comparisons, as they're handled by `useless_comparison`.
+    if matches!(value.node, ExprKind::Compare { .. }) {
+        return;
+    }
+
+    // Ignore strings, to avoid false positives with docstrings.
```
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_bugbear/rules/useless_expression.rs:44]: Adding a small code snipped here of a case where a docstring causes a false positive would be helpful. 

Is there a way to determine whether a string is a docstring?

File: crates/ruff/src/rules/flake8_bugbear/rules/useless_expression.rs [hunk 2]
```diff
@@ -1,41 +1,77 @@
...
             }
+    ) {
+        return;
+    }
+
+    // Ignore statements that have side effects.
+    if contains_effect(&checker.ctx, value) {
+        // Flag attributes as useless expressions, even if they're attached to calls or other
+        // expressions.
+        if matches!(value.node, ExprKind::Attribute { .. }) {
```
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_bugbear/rules/useless_expression.rs:60]: Should we use a different diagnostic in that case?

**Review by MichaReiser — APPROVED** [review]
Could you update your PR summary to include documentation of what you changed: How does this PR fix the mentioned issue and does it fix all issues?

**Comment by henryiii** [comment]
The one in `build` is useless, certainly! :)

**Pull request merged**